src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -42,11 +45,144 @@ public void RegisterReceiver(Receiver r)
 
         private static readonly Telemetry Singleton = new Telemetry();
 
-        private Telemetry(){}
+        internal Telemetry(){}  // This is an internal constructor to build isolated unit test instance
 
         public static Telemetry GetInstance()
         {
             return Singleton;
         }
+
+        public bool TelemetryOnFailureOnly { get; set; }
+
+        private ConcurrentDictionary<Tuple<string, string>, EventBase> EventsInProgress = new ConcurrentDictionary<Tuple<string, string>, EventBase>();
+
+        private ConcurrentDictionary<string, List<EventBase>> CompletedEvents = new ConcurrentDictionary<string, List<EventBase>>();
+
+        internal string GenerateNewRequestId()
+        {
+            return Guid.NewGuid().ToString();
+        }
+
+        internal void StartEvent(string requestId, EventBase eventToStart)
+        {
+            if (requestId != null)
+            {
+                EventsInProgress[new Tuple<string, string>(requestId, eventToStart[EventBase.EventName])] = eventToStart;
+            }
+        }
+
+        internal void StopEvent(string requestId, EventBase eventToStop)
+        {
+            if (requestId == null)
+            {
+                return;
+            }
+            Tuple<string, string> eventKey = new Tuple<string, string>(requestId, eventToStop[EventBase.EventName]);
+
+            // Locate the same name event in the EventsInProgress map
+            EventBase eventStarted = null;
+            if (EventsInProgress.ContainsKey(eventKey))
+            {
+                eventStarted = EventsInProgress[eventKey];
+            }
+
+            // If we did not get anything back from the dictionary, most likely its a bug that StopEvent
+            // was called without a corresponding StartEvent
+            if (null == eventStarted)
+            {
+                // Stop Event called without a corresponding start_event.
+                return;
+            }
+
+            // Set execution time properties on the event
+            eventToStop.Stop();
+
+            if (!CompletedEvents.ContainsKey(requestId))
+            {
+                // if this is the first event associated to this
+                // RequestId we need to initialize a new List to hold
+                // all of sibling events
+                List<EventBase> events = new List<EventBase>();
+                events.Add(eventToStop);
+                CompletedEvents[requestId] = events;
+            }
+            else
+            {
+                // if this event shares a RequestId with other events
+                // just add it to the List
+                CompletedEvents[requestId].Add(eventToStop);
+            }
+
+            // Mark this event as no longer in progress
+            EventBase dummy;
```

Comment:
Yeah, I agree, that looks ugly. We used to have an elegant `EventsInProgress.Remove(eventKey)` when the `EventsInProgress` was a `Dictionary`. Now we switched to use `ConcurrentDictionary` and it turns out it does not provide `Remove()`, at least [not easily](http://stackoverflow.com/questions/7464965/why-does-concurrentdictionary-tryremove-require-a-second-out-argument).